src/cregistry/distfile.c

```diff
@@ -0,0 +1,316 @@
+ * @param [in] keys      a list of keys to search by
+ * @param [in] vals      a list of values to search by, matching keys
+ * @param [in] strats    a list of strategies to use when searching
+ * @param [in] key_count the number of key/value pairs passed
+ * @param [out] distfiles    a list of matching distfiles
+ * @param [out] errPtr   on error, a description of the error that occurred
+ * @return               the number of entries if success; false if failure
+ */
+int reg_distfile_search(reg_registry* reg, char** keys, char** vals, int* strats,
+        int key_count, reg_distfile*** distfiles, reg_error* errPtr) {
+    int i;
+    char* kwd = " WHERE ";
+    char* query;
+    size_t query_len, query_space;
+    int result;
+
+    /* build the query */
+    query = strdup("SELECT id, subdir, path FROM registry.distfiles");
+    if (!query) {
+        return -1;
+    }
+    query_len = query_space = strlen(query);
+
+    for (i = 0; i < key_count; i++) {
+        char* op;
+        char* cond;
+
+        /* get the strategy */
+        if ((op = reg_strategy_op(strats[i], errPtr)) == NULL) {
+            free(query);
+            return -1;
+        }
+
+        cond = sqlite3_mprintf(op, keys[i], vals[i]);
+        if (!cond || !reg_strcat(&query, &query_len, &query_space, kwd)
+            || !reg_strcat(&query, &query_len, &query_space, cond)) {
+            free(query);
+            return -1;
+        }
+        sqlite3_free(cond);
+        kwd = " AND ";
+    }
+
+    /* do the query */
+    result = reg_all_distfiles(reg, query, -1, distfiles, errPtr);
+    free(query);
+    return result;
+}
+
+/**
+ * Gets a named property of a distfile. That property can be set using
+ * `reg_distfile_propset`. The property named must be one that exists in the table
+ * and must not be one with internal meaning such as `id`.
+ *
+ * @param [in] distfile    distfile to get property from
+ * @param [in] key     property to get
+ * @param [out] value  the value of the property
+ * @param [out] errPtr on error, a description of the error that occurred
+ * @return             true if success; false if failure
+ */
+int reg_distfile_propget(reg_distfile* distfile, char* key, char** value,
+        reg_error* errPtr) {
+    reg_registry* reg = distfile->reg;
+    int result = 0;
+    sqlite3_stmt* stmt = NULL;
+    char* query;
+    const char *text;
+    query = sqlite3_mprintf(
+            "SELECT %q FROM registry.distfiles "
+            "WHERE id=%lld AND subdir='%q' AND path='%q'", key,
+            distfile->key.id, distfile->key.subdir, distfile->key.path);
+    if (sqlite3_prepare_v2(reg->db, query, -1, &stmt, NULL) == SQLITE_OK) {
+        int r;
+        do {
+            r = sqlite3_step(stmt);
+            switch (r) {
+                case SQLITE_ROW:
+                    text = (const char*)sqlite3_column_text(stmt, 0);
+                    if (text) {
+                        *value = strdup(text);
```

Comment:
We don't have error handling for this strdup(3) (here or in `entry.c`, `file.c`, `propget.c`). We should probably add some.